File: lib/C110PSerial/ProtoFrame.cpp
#include "ProtoFrame.h"

bool ProtoFrame::readFrame()
{
    uint32_t timestamp = this->getSafeTimestamp();
    while (m_stream->available())
    {
        // std::cout << "[DEBUG] m_inputIndex: " << m_inputIndex << std::endl;
        // std::cout << "[DEBUG] m_inputLength: " << m_inputLength << std::endl;
        // Check elapsed time
        uint32_t now = this->getSafeTimestamp();
        if (now - timestamp > m_messageTimeout) {
            // Exceeded max duration, exit loop
            break;
        }

        // returns an int so that it can return all 255 possible 8 bit codes
        // plus still be able to return a -1 (0xFFFF) to indicate that nothing was actually read
        int8_t c = m_stream->read();
        std::cout << "[DEBUG] Read byte: 0x" << std::hex << static_cast<int>(c) << std::dec << std::endl;
        if (c == -1)
        {
            // No data available
            break;
        }
        if (m_inputIndex == 0 && c == START_BYTE)
        {
            // 
            std::cout << "[DEBUG] Start of new message" << std::endl;
            m_inputIndex = 1;
            continue;
        }
        else if (m_inputIndex == 1)
        {
            // 
            std::cout << "[DEBUG] Second byte should be the length" << std::endl;
            m_inputLength = static_cast<size_t>(c);
            if (m_inputLength > MAX_SIZE - 1)
            {
                // 
                std::cout << "[DEBUG] Invalid length: reset" << std::endl;
                m_inputIndex = 0;
                m_inputLength = 0;
                m_inputCrc = 0;
                // sendNack(0, "Invalid length");
                return false;
            }
            else
            {
                m_inputIndex++;
            }
            continue;
        }
        else if (m_inputIndex == m_inputLength + 2)
        {
            // should be the CRC
            m_inputCrc = static_cast<uint8_t>(c);
            // 
            std::cout << "[DEBUG] verify CRC: received=" << static_cast<int>(m_inputCrc)
                      << ", calculated=" << static_cast<int>(crc8.calculate(m_inputBuffer, m_inputLength)) << std::endl;
            std::cout << "[DEBUG] m_inputBuffer: ";
            for (size_t i = 0; i < m_inputLength; ++i) {
                std::cout << std::hex << std::setw(2) << std::setfill('0')
                          << static_cast<int>(m_inputBuffer[i]) << " ";
            }
            std::cout << std::dec << std::endl;
            if (crc8.calculate(m_inputBuffer, m_inputLength) == m_inputCrc)
            {
                receiveMessage(m_inputBuffer, m_inputLength);
                m_inputIndex = 0; 
                m_inputLength = 0;
                m_inputCrc = 0;
                return true;
            }
            else
            {
                std::cout << "[DEBUG] CRC mismatch: reset" << std::endl;
                m_inputIndex = 0; 
                m_inputLength = 0;
                m_inputCrc = 0;
                return false;
            }
            continue;
        }
        else if (m_inputIndex > 1)
        {
            // Middle of message
            if (m_inputIndex < BUFFER_MESSAGE_MAX_SIZE - 1)
            {
                m_inputBuffer[(m_inputIndex++)-2] = static_cast<uint8_t>(c);
            }
            else
            {
                // 
                std::cout << "[DEBUG] Buffer overflow: reset" << std::endl;
                m_inputIndex = 0;
                // sendNack(0, "Input buffer overflow");
            }
            continue;
        }
    }
    std::cout << "[DEBUG] Exiting readFrame (no complete message)" << std::endl;
    return false;
}

void ProtoFrame::handleAck(uint32_t timestamp)
{
    std::cout << "[DEBUG] handleAck: " << timestamp << std::endl;
    m_messageInfoMap.erase(timestamp);
}

void ProtoFrame::sendAck(uint32_t timestamp)
{
    AckCommand ack = { true };
    C110PCommand msg = C110PCommand_init_default;
    msg.id = timestamp;
    msg.data.ack = ack;
    send(msg);
}

void ProtoFrame::sendNack(uint32_t timestamp, const char* reason /* = "Unknown" */)
{
    AckCommand ack = AckCommand_init_default;
    ack.acknowledged = false;
    strncpy(ack.reason, reason, sizeof(ack.reason) - 1);
    // // Set up the callback to encode the reason string
    // ack.reason.funcs.encode = [](pb_ostream_t *stream, const pb_field_t *field, void * const *arg) -> bool {
    //     const char* str = static_cast<const char*>(*arg);
    //     return pb_encode_tag_for_field(stream, field) &&
    //            pb_encode_string(stream, (const uint8_t*)str, strlen(str));
    // };
    // ack.reason.arg = (void*)reason;
    C110PCommand msg = C110PCommand_init_default;
    msg.id = timestamp;
    msg.data.ack = ack;
    send(msg);
}

void ProtoFrame::handleNack(uint32_t timestamp)
{
    // For now, treat NACK like a retriable failure
    C110PCommand* msg = m_sentMessageBuffer.get(timestamp);
    if (msg && m_messageInfoMap.count(timestamp) == 1)
    {
        resendMessage(*msg);
    }
}

void ProtoFrame::retryMessages()
{
    uint64_t currentTime = this->getSafeTimestamp();
    // Retry unacknowledged messages from the sent buffer
    for (auto& pair : m_messageInfoMap)
    {
        uint32_t timestamp = pair.first;
        C110PCommand* msg = m_sentMessageBuffer.get(timestamp);
        if (msg && (currentTime - pair.second.lastProcessedTimestamp >= m_messageTimeout) && pair.second.retryCount < m_maxRetries)
        {
            std::cout << "[DEBUG] Retrying message with timestamp: " << timestamp << std::endl;
            // Resend the message
            resendMessage(*msg);
        }
        else if (msg && pair.second.retryCount >= m_maxRetries)
        {
            std::cout << "[DEBUG] Max retries reached for message with timestamp: " << timestamp << std::endl;
            m_messageInfoMap.erase(timestamp);
        }
    }
}

void ProtoFrame::resendMessage(C110PCommand& message)
{
    // message.processedTimestamp = this->getSafeTimestamp();
    auto it = m_messageInfoMap.find(message.id);
    if (it != m_messageInfoMap.end()) {
        it->second.lastProcessedTimestamp = this->getSafeTimestamp();
        it->second.retryCount++;
    }
    send(message);
}

void ProtoFrame::receiveMessage(const uint8_t* rawMessage, size_t length)
{
    C110PCommand msg = C110PCommand_init_zero;
    std::cout << "Received message" << std::endl;

    pb_istream_t stream = pb_istream_from_buffer(rawMessage, length);
    if (!pb_decode(&stream, C110PCommand_fields, &msg))
    {
        // sendNack(0, "Protobuf decode failed");
        return;
    }
    
    if (m_receivedMessageBuffer.contains(msg.id))
    {
        // Duplicate message: already processed, just re-ACK
        sendAck(msg.id);
    }
    else
    {
        m_receivedMessageBuffer.add(msg);
        sendAck(msg.id);
        processCallback(msg);
    }
}

void ProtoFrame::processCallback(const C110PCommand& message)
{
    std::cout << "[DEBUG] processCallback: which_data=" << message.which_data << std::endl;
    switch(message.which_data)
    {
        case C110PCommand_ack_tag:
            if (message.data.ack.acknowledged)
            {
                std::cout << "[DEBUG] Received ACK for timestamp: " << message.id << std::endl;
                handleAck(message.id);
            }
            else
            {
                std::cout << "[DEBUG] Received NACK for timestamp: " << message.id << std::endl;
                handleNack(message.id);
            }
            break;
        case C110PCommand_led_tag:
            if (m_LedCallback)
            {
                m_LedCallback(message.data.led);
            }
            break;
        case C110PCommand_move_tag:
            if (m_MovementCallback)
            {
                m_MovementCallback(message.data.move);
            }
            break;
        case C110PCommand_sound_tag:
            if (m_SoundCallback)
            {
                m_SoundCallback(message.data.sound);
            }
            break;
        default:
            // Console.println("Unknown message type");
            break;
    }
}

File: python/lib/C110PSerial/ProtoFrame.py
import time
from .CRC8 import CRC8
from .RingBuffer import RingBuffer
from .proto_encode import encode_command
from .proto_decode import decode_command
from .Logger import logger

class ProtoFrame:
    START_BYTE = 0xAA
    MAX_SIZE = 128
    BUFFER_DATA_MAX_SIZE = 128
    BUFFER_MESSAGE_MAX_SIZE = 256

    def __init__(self, stream, identifier=0, timeout=1000, maxRetries=3):
        self.m_regionId = identifier
        self.m_stream = stream
        self.m_sentMessageBuffer = RingBuffer()
        self.m_receivedMessageBuffer = RingBuffer()
        self.m_messageTimeout = timeout
        self.m_maxRetries = maxRetries
        self.m_messageInfoMap = {}
        self.m_inputBuffer = bytearray(self.BUFFER_MESSAGE_MAX_SIZE)
        self.m_inputIndex = 0
        self.m_inputLength = 0
        self.m_inputCrc = 0
        self.m_timestampProvider = lambda: int(time.time())
        self.m_LedCallback = lambda msg: None
        self.m_SoundCallback = lambda msg: None
        self.m_MoveCallback = lambda msg: None
        self.crc8 = CRC8()

        if hasattr(self.m_stream, 'any'):
            # MicroPython
            self.bytesAvailable = lambda: self.m_stream.any()
        elif hasattr(self.m_stream, 'in_waiting'):
            # CircuitPython
            self.bytesAvailable = lambda: self.m_stream.in_waiting
        elif hasattr(self.m_stream, 'available'):
            # Python
            self.bytesAvailable = lambda: self.m_stream.available()
        else:
            raise ValueError("Stream does not support bytesAvailable method")

    def reset(self):
        self.m_sentMessageBuffer.reset()
        self.m_receivedMessageBuffer.reset()
        self.m_messageInfoMap = {}
        self.m_inputIndex = 0
        self.m_inputLength = 0
        self.m_inputCrc = 0

    def setTimestampProvider(self, provider):
        self.m_timestampProvider = provider

    def getSafeTimestamp(self):
        return self.m_timestampProvider() & 0xFFFFFFFF

    def setLedCallback(self, cb):
        self.m_LedCallback = cb

    def setSoundCallback(self, cb):
        self.m_SoundCallback = cb

    def setMoveCallback(self, cb):
        self.m_MoveCallback = cb

    def send(self, msg):
        self.trackMessage(msg)
        return False
    
    def trackMessage(self, msg):
        self.m_sentMessageBuffer.add(msg)
        # TODO: m_messageInfoMap should be a RingBuffer; if nerver ack, it'll grow unbounded
        if msg["id"] not in self.m_messageInfoMap:
            self.m_messageInfoMap[msg["id"]] = {
                "lastProcessedTimestamp": self.getSafeTimestamp(),
                "retryCount": 0
            }
        else:
            self.m_messageInfoMap[msg["id"]]["lastProcessedTimestamp"] = self.getSafeTimestamp()
            self.m_messageInfoMap[msg["id"]]["retryCount"] += 1

    def receive(self, msg):
        self.m_receivedMessageBuffer.add(msg)
        return True

    def readFrame(self):
        timestamp = self.getSafeTimestamp()
        while self.bytesAvailable():
            now = self.getSafeTimestamp()
            if now - timestamp > self.m_messageTimeout:
                break

            c = self.m_stream.read(1)
            if not c:
                break
            logger.debug(f"Received byte: {c}")

            if self.m_inputIndex == 0 and c == self.START_BYTE:
                self.m_inputIndex = 1
                continue
            elif self.m_inputIndex == 1:
                self.m_inputLength = c
                if self.m_inputLength > self.MAX_SIZE - 1:
                    self.m_inputIndex = 0
                    self.m_inputLength = 0
                    self.m_inputCrc = 0
                    logger.error(f"Invalid length: {self.m_inputLength}")
                    return False
                else:
                    self.m_inputIndex += 1
                continue
            elif self.m_inputIndex == self.m_inputLength + 2:
                self.m_inputCrc = c
                if self.crc8.calculate(self.m_inputBuffer[:self.m_inputLength]) == self.m_inputCrc:
                    self.receiveMessage(self.m_inputBuffer[:self.m_inputLength])
                    self.m_inputIndex = 0
                    self.m_inputLength = 0
                    self.m_inputCrc = 0
                    return True
                else:
                    self.m_inputIndex = 0
                    self.m_inputLength = 0
                    self.m_inputCrc = 0
                    logger.error("Invalid CRC: {m_inputCrc}")
                    return False
            elif self.m_inputIndex > 1:
                if self.m_inputIndex < self.BUFFER_MESSAGE_MAX_SIZE - 1:
                    self.m_inputBuffer[self.m_inputIndex - 2] = c
                    self.m_inputIndex += 1
                else:
                    self.m_inputIndex = 0
                continue
        logger.debug("No data available")
        return False

    def getSentMessageBufferSize(self):
        return len(self.m_sentMessageBuffer)

    def getReceivedMessageBufferSize(self):
        return len(self.m_receivedMessageBuffer)

    def getUnacknowledgedMessagesSize(self):
        return len(self.m_messageInfoMap)

    def getUnacknowledgedMessage(self, timestamp):
        return timestamp if timestamp in self.m_messageInfoMap else 0

    def getLastSentMessage(self):
        return self.m_sentMessageBuffer.getCurrentValue()

    def getLastReceivedMessage(self):
        return self.m_receivedMessageBuffer.getCurrentValue()

    def handleAck(self, timestamp):
        self.m_messageInfoMap.pop(timestamp, None)

    def sendAck(self, timestamp):
        msg = {
            "id": timestamp,
            "ack": {
                "acknowledged": True,
                "reason": None
            }
        }
        self.send(msg)
        self.m_messageInfoMap.pop(timestamp, None)

    def sendNack(self, timestamp, reason="Unknown"):
        msg = {
            "id": timestamp,
            "ack": {
                "acknowledged": False,
                "reason": reason
            }
        }
        self.send(msg)

    def handleNack(self, timestamp):
        if timestamp in self.m_sentMessageBuffer and timestamp in self.m_messageInfoMap:
            self.resendMessage(self.m_sentMessageBuffer[timestamp])

    def retryMessages(self):
        currentTime = self.getSafeTimestamp()
        for timestamp, info in list(self.m_messageInfoMap.items()):
            if (currentTime - info['lastProcessedTimestamp'] >= self.m_messageTimeout and
                info['retryCount'] < self.m_maxRetries):
                self.resendMessage(self.m_sentMessageBuffer[timestamp])
            elif info['retryCount'] >= self.m_maxRetries:
                del self.m_messageInfoMap[timestamp]

    def resendMessage(self, msg):
        self.send(msg)

    def receiveMessage(self, rawMessage):
        # Replace with your own message decoding logic
        err, msg = decode_command(rawMessage)
        logger.debug(f"Decoded message: {msg}")
        if err:
            if msg["id"] != 0:
                self.sendNack(msg["id"], "Invalid message")
            return
        if self.m_receivedMessageBuffer.contains(msg["id"]):
            self.sendAck(msg['id'])
        else:
            self.receive(msg)
            self.sendAck(msg['id'])
            self.processCallback(msg)

    def processCallback(self, message):
        if 'ack' in message:
            if message['ack']['acknowledged']:
                self.handleAck(message['id'])
            else:
                self.handleNack(message['id'])
        elif 'led' in message and self.m_LedCallback:
            self.m_LedCallback(message['led'])
        elif 'move' in message and self.m_MoveCallback:
            self.m_MoveCallback(message['move'])
        elif 'sound' in message and self.m_SoundCallback:
            self.m_SoundCallback(message['sound'])
